### PR TITLE
feat(routing): add classifier router for dynamic model and tool selection

### PR DESCRIPTION
**Pull request by pdogra1299**
## Summary

Adds an opt-in **Classifier Router**: NeuroLink classifies each request and selects a provider/model from a configurable "available base" pool — routing hard tasks to more capable models and easy tasks to cheaper/faster ones — and can narrow the tool set per request. Disabled by default, fails open, fully backward compatible.

## What's included

- **Types** — `src/lib/types/classifierRouter.ts` (`Classifier*`), plus `classifierRouter` on the constructor config.
- **Strategies** — `src/lib/routing/classifierStrategies.ts`:
  - `classifyHeuristic` — zero-cost, 5-tier generalization of the existing task scorer (pure prompt-text, provider-agnostic).
  - `classifyLlm` — a cheap classifier model returns a difficulty and can pick a model **directly** from described candidates.
- **Router** — `src/lib/routing/classifierRouter.ts`: difficulty → model via LLM direct pick / `tierMap` / per-member `tiers` / `cost`+`quality` scoring (auto-enriched from the model registry), with capability filtering and per-difficulty tool directives.
- **Wiring** — `neurolink.ts` `generate()`/`stream()` pre-call seam (before `requestRouter`). Precedence: caller-pins > classifierRouter > requestRouter > orchestration; stands down when a `modelPool` is configured; the LLM classifier's own call can't recurse (guarded via `context.__classifierRouted`).
- **CLI flags** — `--classifier-router`, `--classifier-strategy`, `--classifier-model-provider/-name/-region`, `--classifier-pool` (file or inline JSON), `--classifier-timeout` (threaded via `globalSessionState`, mirroring `--tool-routing-*`).
- **Docs** — `docs/features/classifier-router.md` (+ pointer from the features index).

## Generic for custom models (LiteLLM / OpenAI-compatible / Ollama)

No registry dependency: route via per-member `tiers`/`tierMap`, or let the LLM classifier pick by plain-English `description`.

## Testing

- New suite `test/continuous-test-suite-classifier-router.ts` — **19 cases** (heuristic tiers, metadata scoring, `tierMap`, capability filter, tool directives, LLM direct-pick + invalid/absent fallbacks). Run: `pnpm run test:classifier-router`.
- `pnpm run check` (tsc strict) clean · `lint` 0 errors · full `build` 0 errors / publint OK.
- Live (Vertex): easy prompt → `gemini-2.5-flash`, hard prompt → `gemini-2.5-pro`, verified via both SDK and CLI.

## Backward compatibility

Opt-in (`enabled: false` by default); a default `new NeuroLink()` is byte-for-byte unchanged; all new config/exports are additive.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added an opt-in Classifier Router that can automatically route requests to a more suitable model based on request difficulty.
  * Expanded CLI support so the router can be enabled and configured from the command line.
  * Added support for per-difficulty tool selection and model-pool routing options.

* **Documentation**
  * Added a new guide for the Classifier Router and updated the product docs index with a reference to it.

* **Tests**
  * Added coverage for routing behavior, fallback handling, and tool-scoping scenarios.

### CODE REVIEW

File: src/lib/routing/classifierStrategies.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+  generate: ClassifierGenerateFn,
+  classifierModel?: ClassifierModelRef,
+  timeoutMs?: number,
+  candidates?: ClassifierCandidate[],
+): Promise<ClassifierDecision> {
+  const lines = [
+    "Task to classify:",
+    '"""',
+    (input.prompt ?? "").slice(0, 4000),
+    '"""',
```
Comment (Tara-ag) [inline src/lib/routing/classifierStrategies.ts:131]: 💡 **MINOR: Prompt content may be logged/sent without sensitivity check**

The `classifyLlm` function sends the user prompt to the classifier model:

```typescript
(input.prompt ?? "").slice(0, 4000),
```

While this is limited to 4000 chars, there's no check for whether the prompt might contain sensitive data (PII, credentials, etc.) before sending it to the classifier LLM. This could inadvertently leak sensitive information in:
- The classifier LLM's logs
- The classifier LLM's training data (if the provider retains queries)

**Suggestion:** Add a warning in the documentation about this behavior, or consider adding an optional `sanitizePrompt` hook for users who need to strip sensitive patterns before classification.

See bootstrapped-standards: "Never log raw user content (transcripts, tool arguments, router output) at INFO level; use metadata-only logging or guard with logger.shouldLog('debug')."

File: src/lib/neurolink.ts [hunk 2]
```diff
@@ -4435,6 +4477,128 @@ Current user's request: ${currentInput}`;
     }
```
Comment (Tara-ag) [inline src/lib/neurolink.ts:4477]: ✅ **Good practice: Proper precedence handling**

The classifier router integration correctly respects the precedence order documented in the PR:

```typescript
// caller-pinned provider+model > classifierRouter > requestRouter > orchestration
```

The early return when `__classifierRouted` is set (line 4480) and the check for `modelPool` (line 4487) ensure proper layering. This follows the established pattern from `applyRequestRouter` and maintains backward compatibility.

File: src/lib/routing/classifierRouter.ts [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+export class ClassifierRouter {
+  private readonly metaCache = new Map<string, ClassifierModelMeta>();
+
+  constructor(
+    private readonly config: ClassifierRouterConfig,
+    private readonly deps: ClassifierRouterDeps = {},
+  ) {}
+
+  /**
+   * Classify the request and produce a combined model + tool decision, or
```
Comment (Tara-ag) [inline src/lib/routing/classifierRouter.ts:62]: ⚠️ **MAJOR: Unbounded cache growth potential**

The `metaCache` Map stores metadata per (provider, model) combination without any size limit or eviction policy:

```typescript
private readonly metaCache = new Map<string, ClassifierModelMeta>();
```

In long-running processes or when using dynamic model pools, this cache could grow unbounded over time. Consider adding:
- A maximum size limit with LRU eviction
- Periodic cleanup of stale entries
- Or document that this is intentionally unbounded with the assumption that pool configurations are relatively static

**Suggested fix:**
```typescript
// Add near the cache declaration
private readonly MAX_META_CACHE_SIZE = 1000;

// In metaFor(), before set()
if (this.metaCache.size >= this.MAX_META_CACHE_SIZE) {
  // Evict oldest entry (or use LRU)
  const firstKey = this.metaCache.keys().next().value;
  this.metaCache.delete(firstKey);
}
```

File: src/cli/utils/classifierRouterFlags.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+      ...(hasProvider && {
+        provider: flags.classifierModelProvider as string,
+      }),
+      ...(hasModel && { model: flags.classifierModelName as string }),
+      ...(hasRegion && { region: flags.classifierModelRegion as string }),
+    };
+  }
+
+  return config;
+}
```
Comment (Tara-ag) [inline src/cli/utils/classifierRouterFlags.ts:91]: ⚠️ **MAJOR: Path traversal vulnerability in pool file loading**

The `parsePoolFlag` function uses user-provided input directly with filesystem operations:

```typescript
const resolved = fs.existsSync(value) ? value : null;
// ...
const jsonText = resolved ? fs.readFileSync(resolved, "utf8") : value;
```

While this is a CLI flag (user-supplied), there's no validation to prevent path traversal attacks like `--classifier-pool ../../../etc/passwd`. The size check only applies after the file is resolved.

**Suggested fix:**
```typescript
import path from "node:path";

function parsePoolFlag(value: string, cwd: string = process.cwd()): ClassifierRouterPoolMember[] {
  // Resolve relative to cwd and ensure it stays within bounds
  const resolved = path.resolve(cwd, value);
  if (!resolved.startsWith(cwd)) {
    logger.warn("[classifier-router] Pool file path escapes working directory; ignoring.");
    return [];
  }
  // ... rest of function
}
```

File: src/lib/routing/classifierStrategies.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+      if (c.capabilities && c.capabilities.length > 0) {
+        bits.push(`caps: ${c.capabilities.join(",")}`);
+      }
+      lines.push(`- id="${c.id}": ${bits.join(" — ")}`);
+    }
+    lines.push("", "Set selectedModelId to the chosen id (omit if unsure).");
+  }
+
+  const result = await generate({
+    input: { text: lines.join("\n") },
```
Comment (Tara-ag) [inline src/lib/routing/classifierStrategies.ts]: 💡 **MINOR: LLM classifier timeout relies on underlying generate() implementation**

The `classifyLlm` function passes a timeout to the generate call:

```typescript
timeout: timeoutMs ?? 8000,
```

However, this timeout is only effective if the injected `generate` function properly respects and implements the timeout. If the underlying `NeuroLink.generate()` doesn't properly enforce timeouts (e.g., during network stalls), the classifier could hang indefinitely.

**Suggestion:** Consider wrapping the generate call with `withTimeout()` as a defensive measure, per CLAUDE.md error handling guidelines:

```typescript
const result = await withTimeout(
  generate({ ... }),
  timeoutMs ?? 8000,
  new Error("Classifier LLM call timed out")
);
```

**Review by Tara-ag — COMMENTED** [review]
## Review Summary: Classifier Router Feature

I've reviewed the Classifier Router PR (#1119) which adds an opt-in request classification and routing system. This is a substantial feature with 1,813 additions across 16 files.

### Issues Found

| Severity | Count | Description |
|----------|-------|-------------|
| MAJOR | 2 | Unbounded cache growth, Path traversal vulnerability |
| MINOR | 2 | Sensitive data handling, Timeout enforcement |

### Detailed Findings

**MAJOR Issues:**

1. **Unbounded cache in `ClassifierRouter`** (`src/lib/routing/classifierRouter.ts:56`)
   - The `metaCache` Map has no size limit or eviction policy
   - Could cause memory issues in long-running processes with dynamic pools
   - **Suggested fix:** Add LRU eviction with a reasonable max size (e.g., 1000 entries)

2. **Path traversal vulnerability** (`src/cli/utils/classifierRouterFlags.ts:85`)
   - User-provided `--classifier-pool` path is used directly with filesystem operations
   - No validation prevents paths like `../../../etc/passwd`
   - **Suggested fix:** Resolve and validate paths stay within working directory

**MINOR Issues:**

3. **Sensitive data in classifier prompts** (`src/lib/routing/classifierStrategies.ts:130`)
   - User prompts are sent to the classifier LLM without sanitization
   - Could leak PII/credentials to the classifier model's logs
   - **Suggested fix:** Document this behavior or add optional sanitization hook

4. **Timeout reliance on generate() implementation** (`src/lib/routing/classifierStrategies.ts:155`)
   - The timeout is passed but not defensively enforced
   - **Suggested fix:** Wrap with `withTimeout()` for guaranteed enforcement

### Positive Observations

✅ **Backward compatibility** - Feature is opt-in (`enabled: false` by default)  
✅ **Fail-open design** - Any classifier error leaves the call unchanged  
✅ **Proper precedence** - Respects caller-pinned > classifier > requestRouter > orchestration  
✅ **Good test coverage** - 19 test cases covering heuristic, LLM, and edge cases  
✅ **Documentation** - Comprehensive docs with examples  
✅ **No breaking changes** - All new config/exports are additive

### Recommendation

**Approve with minor revisions.** The MAJOR issues should be addressed before merging:
1. Add cache size limiting
2. Add path traversal protection

The feature is well-designed, thoroughly tested, and maintains backward compatibility. The architectural integration follows existing patterns in the codebase.

**Comment by pdogra1299** [comment]
Addressed the review feedback in the amended commit (single-commit policy preserved).

### CI build fix — `Validate Documentation` (was failing)
Root cause: broken doc links. `sync-docs` rewrites relative `.md` links and the Docusaurus build (`onBrokenLinks: "throw"`) rejected the results (`/docs/provider-fallback`, `/docs/per-request-credentials`, `/docs/classifier-router`). Switched the new doc and the features-index link to the repo's absolute `/docs/features/...` convention (matching `credential-validation.md`). Reproduced the Docusaurus build locally — now passes with **0 broken links**.

### Review comments
- **`classifierRouter.ts` — unbounded `metaCache` (MAJOR):** added a 1000-entry cap with FIFO eviction.
- **`classifierRouterFlags.ts` — pool-file path traversal (MAJOR):** documented the trust model (a local, user-supplied CLI flag — identical to the existing `--tool-routing-servers`; absolute/home paths are intentionally supported and the input is size-capped). There is no cross-trust boundary to confine here, and cwd-confinement would break the documented absolute-path support.
- **`classifierStrategies.ts` — LLM timeout reliance (MINOR):** wrapped the classifier `generate()` in `withTimeout()` for a hard wall-clock ceiling; the `TimeoutError` fails open to the heuristic.
- **`classifierStrategies.ts` — prompt sensitivity (MINOR):** added a "Prompt privacy" caveat to the docs; the default `heuristic` strategy keeps classification fully in-process (no prompt leaves the environment).
- **docs `classifierModel` example (CodeRabbit):** fixed to valid TypeScript with explicit field types.

Verified locally: `tsc --strict` clean · ESLint clean on changed files · `pnpm build` (SDK) green · `docs-site` build green · 19/19 classifier-router suite.

**Comment by Tara-ag** [comment]
## Review Summary: Classifier Router Feature

I've completed my review of PR #1119 which adds the Classifier Router feature. This is a well-structured feature addition with good test coverage and documentation. However, I've identified **4 MAJOR issues** that should be addressed before merging:

### 🔒 Security Issues (MAJOR)

1. **Path traversal vulnerability** (`src/cli/utils/classifierRouterFlags.ts`): The `parsePoolFlag` function doesn't validate that the resolved path stays within the working directory, allowing potential access to sensitive files via `--classifier-pool ../../../etc/passwd`.

### 🏗️ Architecture/Correctness Issues (MAJOR)

2. **Unbounded cache growth** (`src/lib/routing/classifierRouter.ts`): The `metaCache` Map has no size limit or eviction policy, which could cause memory issues in long-running processes.

3. **Cache key collision** (`src/lib/routing/classifierRouter.ts`): Pool members with the same provider/model but different declared `cost`/`quality`/`capabilities` share the same cache key, causing incorrect metadata to be used for ranking.

4. **Stale config in long-lived CLI** (`src/cli/utils/classifierRouterFlags.ts`): Returning `undefined` when `--classifier-router` is absent leaves previously stored config in `globalSession`, causing commands to inherit stale routing configuration.

### 💡 Minor Suggestions

- **PII leakage risk** (`src/lib/routing/classifierStrategies.ts`): The LLM classifier sends user prompts to the classifier model without sensitivity checks - consider documenting this risk.
- **Test coverage gap** (`package.json`): The new `test:classifier-router` script isn't included in the default `test:unit` flow, risking regression gaps.

### ✅ What's Good

- Comprehensive test suite (19 test cases)
- Good documentation with clear examples
- Proper fail-open design with fallback to heuristic
- Clean separation of concerns between strategies and router
- Backward compatible (opt-in, disabled by default)

### Next Steps

Please address the 4 MAJOR issues above. The path traversal and cache key collision issues are particularly important as they could lead to security vulnerabilities or incorrect routing behavior.

---

*Note: I've posted inline comments on the specific files for each issue.*
